tests/include/test/helpers.h

```diff
@@ -53,7 +53,29 @@
 int mbedtls_test_platform_setup( void );
 void mbedtls_test_platform_teardown( void );
 
-int mbedtls_test_unhexify( unsigned char *obuf, const char *ibuf );
+/**
+ * \brief          This function translates an ASCII string encoding an
+ *                 hexadecimal number into the encoded hexadecimal number. The
+ *                 hexadecimal number is represented as an array of
+ *                 unsigned char.
+ *
+ * \note           The output buffer can be the same as the input buffer. For
+ *                 any other overlapping of the input and output buffers, the
+ *                 behavior is undefined.
+ *
+ * \param obuf     Output buffer.
+ * \param obufmax  Size in number of bytes of \p obuf.
+ * \param ibuf     Input buffer.
+ * \param len      The number of unsigned char written in \p obuf. This must
+ *                 not be \c NULL.
+ *
+ * \return         \c 0 on success.
+ * \return         \c -1 if the output buffer is too small or the input string
+ *                 is not a valid ASCII encoding of an hexadecimal number.
+ */
+int mbedtls_test_unhexify( unsigned char *obuf, size_t obufmax,
+                           const char *ibuf, size_t *len );
```

Comment:
I am happy to backport this commit to LTS branches. Regarding changing `char*` to `data_t` I don't see that fitting well in this PR.

Reply:
Many of the changes in this PR would not have been necessary if the code used `data_t` where possible. It's unrelated to the main goal of the PR, but so is the change of interface of `mbedtls_test_unhexify`.

Reply:
You mean the changes in this commit?

Reply:
Background: up to Mbed TLS 2.12, the unit test framework allowed test functions to have arguments of type either `int` or `char *` (null-terminated string). A lot of unit tests received binary data encoded a hex string, which the test function had to decode (unhexify) into a separate buffer. The new unit test generation script introduced in Mbed TLS 2.13 added a third possibility, `data_t*`, with the same hex string syntax in the `.data` file, but where the test framework would take care of the unhexification so that the test function could operate directly on the binary data.

With the availability of `data_t*` test function parameters, there is very rarely a need for hex strings in arguments. When we added `data_t*`, we also converted most of the test function code that took a hex string as argument and unhexified it to take a `data_t*` instead. However some instances got missed, either because they used the unhexification idiom slightly differently, or because they were in flight in parallel with the introduction of the new unit test framework, or because they were plain forgotten. We never took the time to hunt for these stragglers.

In this PR, I see a lot of changes where `mbedtls_test_unhexify` is called in a case that, ideally, should have been converted to a use of `data_t*` in 2.13: where the hex string is an argument to the test function that is solely used to be decoded from hex. Ideally those calls to `mbedtls_test_unhexify` should no longer exist.

Now we have a choice between:

* Change these calls to the new unhexify syntax in development. Then, each time we backport to 2.16, we have to convert to the old unhexify syntax.
* Backport the new unhexify syntax (function definition, and all the calls).
* Get rid of most of the unhexify calls by using `data_t*` instead, which makes the backport difficulty moot.

The reason not to replace obsolescent unhexify calls by `data_t*` arguments was to avoid hunting for them and changing them. But this PR goes and changes all of these unhexify calls anyway. Changing them to `data_t*` parameters would be about the same amount of work to code, and would have been significantly easier to review (though it's a bit late for that).

I chose to review the unhexify changes and not to reject them, but I feel we're throwing good money after bad: the largest change in this PR is to function calls that don't need to be there at all.

Reply:
Thanks for the very good explanation. Now it makes completely sense to me to get rid of the unhexify calls we can get rid of by using `data_t*`. Working on it.

Reply:
I've been able to remove all the calls to mbedtls_test_unhexify() in the unit test code.

Reply:
Thank you for doing the rewrite! I think it makes the code more readable and maintainable. I like the way you structured the commits, too, it made the code as easy to review as one could hope.